### PR TITLE
geo/geomfn: Implement ST_SwapOrdinates

### LINKED ISSUE #49050: geo/geomfn: implement ST_SwapOrdinates({geometry,cstring})
Implement `ST_SwapOrdinates` on arguments {geometry,cstring}, which should adopt [PostGIS behaviour](https://postgis.net/docs/ST_SwapOrdinates.html).

*Observers: Please react to this issue if you need this functionality.*

For Geometry builtins, please do the following:
* Ideally add a relevant helper function in [`pkg/geo/geomfn`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo/geomfn) (parse and output related functions can go in [`pkg/geo`](https://github.com/cockroachdb/cockroach/tree/master/pkg/geo)). Add exhaustive unit tests here - you can run through example test cases and make sure that PostGIS and CRDB return the same result within a degree of accuracy (1cm for geography).
  * When using GEOS, you can reference the [C API](https://github.com/libgeos/geos/blob/master/capi/geos_c.h.in) for which functions are available. Unfortunately, Windows is not currently supported when using GEOS.
* Create a new builtin that references this function in [`pkg/sql/sem/builtins/geo_builtins.go`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sem/builtins/geo_builtins.go). Note that we currently do not support optional arguments, so we define functions that have optional arguments once without the optional argument (using the default value in the optional position), and once with the optional argument.
* Modify the tests in [`pkg/sql/logictest/testdata/logic_test/geospatial`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/logictest/testdata/logic_test/geospatial) to call this functionality at least once. You can call `make testbaselogic FILES='geospatial' TESTFLAGS='-rewrite'` to regenerate the output. Tests here should just ensure the builtin is linked end to end (your exhaustive unit tests go the above mentioned packages!).
* Ensure the documentation is regenerated by calling `make buildshort`. You can also play with it by calling `./cockroach demo --empty` afterwards.
* Submit your PR - make sure to follow the guidelines from [creating your first PR](https://wiki.crdb.io/wiki/spaces/CRDB/pages/181633464/Your+first+CockroachDB+PR]).

You can follow #48552 for an example PR.

The following additional guidance has been issued on implementing this function:

re-use code from FlipCoordinates.

<sub>:robot: This issue was synced with a spreadsheet by [gsheets-to-github-issues](https://github.com/cockroachlabs/gsheet-to-github-issues) by [otan](https://github.com/otan) on 2020-09-04T18:54:51Z. Changes to titles, body and labels may be overwritten.</sub>

### PR DESCRIPTION
**Pull request by adzimzf**
Currently PostGIS has the ST_SwapOrdinates, but CRDB hasn't.

Fixes #49050.

Release note (sql change): Implemented the geometry based builtins
`ST_SwapOrdinates`.

### CODE REVIEW

File: pkg/sql/sem/builtins/geo_builtins.go [hunk 1]
```diff
@@ -4333,6 +4333,38 @@ Bottom Left.`,
...
+	"st_swapordinates": makeBuiltin(
+		defProps(),
+		tree.Overload{
+			Types: tree.ArgTypes{
+				{
+					Name: "geometry",
+					Typ:  types.Geometry,
+				},
+				{
+					Name: "cstring",
```
Comment (otan) [inline pkg/sql/sem/builtins/geo_builtins.go:4345]: nit: rename this to `swap_ordinate_string`.

File: pkg/sql/sem/builtins/geo_builtins.go [hunk 2]
```diff
@@ -4333,6 +4333,38 @@ Bottom Left.`,
...
+				ret, err := geomfn.SwapOrdinates(g.Geometry, string(cString))
+				if err != nil {
+					return nil, err
+				}
+
+				return tree.NewDGeometry(ret), nil
+			},
+			Info: infoBuilder{
+				info: `Returns a version of the given geometry with given ordinates swapped.
+The ords parameter is a 2-characters string naming the ordinates to swap. Valid names are: x,y,z and m.`,
```
Comment (otan) [inline pkg/sql/sem/builtins/geo_builtins.go:4363]: nit: space after commas, i.e. `x, y, z and m`

File: pkg/geo/geomfn/swap_ordinates.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	var ordsIndex [2]int
+	for i := 0; i < len(ords); i++ {
+		o, ok := ordsMap[ords[i]]
+		if !ok {
+			return ordsIndex, errInvalidOrdinate
+		}
+		if o > len(src)/l.Stride() {
+			return ordsIndex, fmt.Errorf("geometry does not have an %s ordinate", string(ords[i]))
+		}
+		ordsIndex[i] = o
```
Comment (otan) [inline pkg/geo/geomfn/swap_ordinates.go:73]: the `m` ordinate can be index 2, depending on `geom.Layout`. 
you need to use `l.MIndex()` and `l.ZIndex()` for this to be correct. you can unit test this with a `geom.T` object as long as you test only ingress and outgress `geom.T` objects.

**Review by otan — COMMENTED** [review]
thanks! this is close, a few nits and you should be good to go.

File: pkg/geo/geomfn/swap_ordinates.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	}
+
+	t, err := geometry.AsGeomT()
+	if err != nil {
+		return geometry, err
+	}
+
+	newT, err := applyOnCoordsForGeomT(t, func(l geom.Layout, dst, src []float64) error {
+		if len(ords) != 2 {
+			return errInvalidOrdinate
```
Comment (otan) [inline pkg/geo/geomfn/swap_ordinates.go:36]: nit also: prefer `return errors.New(....)` here, so we get the stack trace.

**Comment by adzimzf** [comment]
Hi @otan,
I have some updates

File: pkg/geo/geomfn/swap_ordinates_test.go [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+			require.True(t, ok)
+
+			require.Equal(t, tt.want, gotCollection)
+		})
+	}
+}
+
+// Test_getOrdsIndices tests the index position of Z and M
+// since currently the geo.Geometry doesn't support ordinates Z and M
+func Test_getOrdsIndices(t *testing.T) {
```
Comment (otan) [inline pkg/geo/geomfn/swap_ordinates_test.go:187]: Rename this to `TestGetOrdsIndicies` - we do not use `_` characters in tests.

File: pkg/geo/geomfn/swap_ordinates_test.go [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+
+			geometryWant, err := geo.MakeGeometryFromGeomT(tt.want)
+			require.NoError(t, err)
+			require.Equal(t, geometryWant, got)
+		})
+	}
+}
+
+func TestSwapOrdinates_Collection(t *testing.T) {
+	tests := []struct {
```
Comment (otan) [inline pkg/geo/geomfn/swap_ordinates_test.go:126]: we normally don't separate test cases like this. this seems easy to fold with the above test cases (i.e. reuse the same struct)? *geom.Collection can be used as geom.T, so you can just extend the number of inputs.

File: pkg/geo/geomfn/swap_ordinates_test.go [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+	"github.com/cockroachdb/cockroach/pkg/geo"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/twpayne/go-geom"
+)
+
+func TestSwapOrdinates(t *testing.T) {
+	tests := []struct {
+		name      string
+		input1    geom.T
```
Comment (otan) [inline pkg/geo/geomfn/swap_ordinates_test.go:25]: can you rename input1 / input2 to `inputGeom` and `inputSwapOrdinates`?

File: pkg/sql/sem/builtins/geo_builtins.go [hunk 8]
```diff
@@ -4333,6 +4333,38 @@ Bottom Left.`,
...
+				ret, err := geomfn.SwapOrdinates(g.Geometry, string(cString))
+				if err != nil {
+					return nil, err
+				}
+
+				return tree.NewDGeometry(ret), nil
+			},
+			Info: infoBuilder{
+				info: `Returns a version of the given geometry with given ordinates swapped.
+The ords parameter is a 2-characters string naming the ordinates to swap. Valid names are: x, y, z and m.`,
```
Comment (otan) [inline pkg/sql/sem/builtins/geo_builtins.go:4363]: nit: if we're saying `ords parameter` here, then we should rename the argument in "swap_ordinate_string" to "ords" - or rename `ords` here to `swap_ordinate_string`

**Review by otan — COMMENTED** [review]
Thanks for the update! Requesting some minor updates to tests.

Also looks like CI is not passing. Can you run `make buildshort` after changing the info string?

**Comment by adzimzf** [comment]
Hi @otan,
I have some new updates,

**Review by otan — APPROVED** [review]

**Comment by otan** [comment]
Thanks for doing this!

bors r+

**Pull request merged**